Basic classes for Query data model

fpacifici opened the pull request:
The first classes needed to provide the Snuba query class with a proper AST.
This PR represents the Query as a tree. The root is the query. Nodes can be containers
(like selected columns), expressions (simple or nested) or conditions (simple or nested).

The basic idea is to provide three ways to manipulate the Query:
- One is a simple iterator to go through all the nodes of a given type (like expressions). This is meant to do whatever we do today with get_all_referenced_columns
- The second is mapping. We would run the mapping function over all the nodes of a given type to replace them (think about replacing columns)
- [This is not implemented yet] Directly navigate the tree from the root (helped with a find() method to find patterns) to provide additional flexibility.

Limitations (will come in followups - see the size of this PR):
- Parsing logic is not there yet
- The top level query navigation methods are not implemented yet 
- formatting logic is not implemented yet

Some notes inline, but as an advice for review:
- start from the query.py file. Then move to nodes.py, then expressions.py
- refer to the tests for examples on how to manipulate the expressions.

fpacifici committed:
Basic classes for Query data model

fpacifici commented on snuba/query/nodes.py:
```diff
@@ -0,0 +1,189 @@
+        raise NotImplementedError
+
+    @abstractmethod
+    def map(self, closure: Callable[[Node], Node]) -> Node:
```
map and filter are a bit of a bet. I think they can be very flexible as a model to edit parts of the query and they can be extended easily to a model where we can do pattern matching of subtrees and replacing them.  On the other end the typing may not be optimal.

fpacifici committed:
Initialize Query class with new data model

fpacifici committed:
Make map and filter have better typing

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
```
Look at the query.py file first, then nodes.py then anything else.

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,123 @@
+    This can be a simple column or a nested expression, but not a condition.
+    """
+    @abstractmethod
+    def map(self, closure: Callable[[Expression], Expression]) -> Expression:
```
I tried a few ways to put map and filter in a generic Node interface but I don't think there was a reasonable solution to make the type checker enforce that the return value of the closure would always be a safe replacement.
Example:
1) if these were defined without generics at Node level, the closure would return a Node, which means there would be no way to make the type checker ensure that a closure applied to an Expression would not return a Condition
2) I tried to have a SubTree class that was generic with regards to the Node type, but the result is that Expression would have to materialize that parameter with itself: class Expresison[SubTree[Expression]] (which is impossible).

The main issue is that this is a tree structure, thus every edit operation that traverses the tree need to return something of the same type of self.

fpacifici commented on snuba/query/query.py:
```diff
@@ -47,11 +52,38 @@ class Query:
+        groupby: Optional[Sequence[Expression]] = None,
+        having: Optional[Sequence[NodeCondition]] = None,
+        order_by: Optional[Sequence[Expression]] = None,
+    ):
```
My rollout strategy is to gradually swap the old state with the new one, piece by piece, when swapping a part of the query, the return types of the public methods of this class will change.
Unfortunately build a second Query class besides this is not very helpful since the accessor methods would return totally different types, so we would still need a wrapper around the new query to change the types.
Swapping a the state piece by piece, makes this class a bit messy in the process, but allows us to test the new model in prod without waiting for the new implementation to be complete.

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Implement Expressions

fpacifici committed:
Support iteration over conditions

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Fix a commnet

fpacifici committed:
Remove the filter method

fpacifici committed:
Introduce Node Collections

fpacifici committed:
Improve test coverage

fpacifici committed:
Add TODOs

fpacifici committed:
Turn some classes into dataclasses

fpacifici committed:
Polishing

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,157 @@
+
+    """
+    @abstractmethod
+    def get_expressions(self) -> ExpressionContainer:
```
This is the trickiest part of this tree.
Here we have a subtree that represent conditions for several levels in the tree. Then at a given level we have Condition leaves (BasicCondition) which on their own are the root of expression trees. So it is not practical (nor useful) to map across both sections of the tree. The only common type is Node. What is more useful is being able to iterate over Expressions across the tree no matter how nested they are, and this is what the ExpressionCOntainer does.

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,134 @@
+class Expression(AliasedNode, ABC):
+    """
+    Abstract representation of a Query node that can be evaluated to a single value.
+    This can be a simple column, NULL or a nested expression, but not a condition.
```
Technically according to many languages grammar, conditions are expression. I think this is a matter of naming, a condition cannot extend Expression as it is (since expresison can have an alias). I could add another level above the two but it does not seem useful since it would not be used anywhere.
(I am implying we are not supporting something like: SELECT a = b or c FROM sentry for now)

tkaemming replied:
I don't really understand why this distinction needs to get made at this level. The functionality that can be expressed (no pun intended) with an expression is a superset of the functionality of a condition, so why not just use the same type internally to represent them?

fpacifici replied:
The only difference right now is that Condition should not be an AliasedNode. but a FormattableNode, while an expression can be assigned an alias.
So Condition cannot just inherit from Expression.
In order to make Condition an Expression we would need :
```
class Expression:

class AliasedExpression(Expression, AliasedNode)

class Column(AliasedExpression)

class Condition(Expression)
```

This way the Expression class does not actually do anything. 
If the naming is confusing I am ok with dropping the term Expression and just having the AliasedExpression. What do you think ?

fpacifici replied:
This would still be a valid point if I turned AliasedNoe into a wrapper, since I would not want AliasedNode to be instantiated with a condition, which could become something like this when formatted:
```
SELECT ... WHERE (a = b) as c
```

fpacifici replied:
This should not be relevant anymore since now a Condition is also an Expression

tkaemming reviewed (commented):
Haven't had a chance to really digest a lot of this, but here are some initial scattershot thoughts.

The biggest question/confusion I have is around why the distinction between expression and condition is necessary (and I suppose aggregation falls into this category as well.)

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,54 @@
+        raise NotImplementedError
+
+
+class CompositeNodeContainer(Generic[TNode], NodeContainer[TNode]):
```
Good question. I assumed I would have to specify this was generic, but indeed I did not try to remove it. Will check and then get back to this.

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,54 @@
+    def __iter__(self) -> Iterator[TNode]:
+        for c in self.__containers:
+            for element in c:
+                yield element
```
I need this to be a NodeContainer (thus with iterable), though since it is not used, it is hard to grasp the reasoning.
I will remove this class for now and reintroduce it when I will be using it.

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,157 @@
+    NOT_IN = "NOT IN"
+    IS = "IS"
+    LIKE = "LIKE"
+    NOT_LIKE = "NOT LIKE"
```
Do you mean having operators as strings ? I don't have a strong opinion on that. It seemed marghinally covenient for query processors to have an enum to rely on, but it is a minor change.

tkaemming replied:
I don't really have a strong opinion one way or another to be honest — was mostly just curious.

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,157 @@
+    NOT_LIKE = "NOT LIKE"
+
+
+class Condition(FormattableNode, ABC):
```
right, I forgot to remove it. In e previous version get_expressions was not abstract.

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,157 @@
+    """
+    This class exists to be able to preserve the NodeContainer[Condition]
+    type at runtime.
+    """
```
that I cannot do 
```
isinstance(a, NodeContainer[Condition])
```
(Actually I am assuming that, I should verify)

but I can do
```
isinstance(a, ConditionContainer)
```

fpacifici commented on snuba/query/nodes.py:
```diff
@@ -0,0 +1,54 @@
+    """
+    A Node in the query tree that knows how to format itself for the clickhouse
+    query.
+    """
```
When the container was itself a Node (previous commit). This is not relevant anymore, I will merge them.

fpacifici commented on snuba/query/nodes.py:
```diff
@@ -0,0 +1,54 @@
+
+    def format(self) -> str:
+        # TODO: do something and call _format_impl
+        raise NotImplementedError
```
Let me try to see how iterating and mapping looks like when an aliased node is actually wrapping the real node.

fpacifici replied:
turned it into a wrapper. Seems working

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,54 @@
+TNode = TypeVar("TNode")
+
+
+class NodeContainer(Generic[TNode], Iterable[TNode], ABC):
```
Good question, let me see what happens if I remove it.

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Fix conditions test

fpacifici committed:
Temporarily remove CompositeNodeContainer

fpacifici committed:
Remove redundant inheritance

fpacifici committed:
Add missing docstring

fpacifici committed:
Rename map to transform

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,54 @@
+from typing import Callable, Generic, Iterable, Iterator, TypeVar
+
+
+TNode = TypeVar("TNode")
```
I removed CompositeContainer for now.

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,54 @@
+    @abstractmethod
+    def map(self, func: Callable[[TNode], TNode]) -> None:
+        """
+        Maps the content of the container in place.
```
replaced with transform

fpacifici commented on snuba/query/nodes.py:
```diff
@@ -0,0 +1,46 @@
+
+
+@dataclass
+class AliasedNode(Node):
```
This as of now does not typecheck because mypy does not like abstract dataclasses, but apparently this is going to be fixed on mypy side:
https://github.com/python/mypy/issues/5374

fpacifici committed:
Make docstring more clear on conditions

fpacifici committed:
Make aliases a wrapper

fpacifici committed:
Fix alias documentation

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Introduce Term

fpacifici committed:
Some documentation improvements

fpacifici committed:
Remove Term abstraction. It does not work

fpacifici committed:
Get rid of sequences of conditions

lynnagara commented on tests/query/test_expressions.py:
```diff
@@ -0,0 +1,131 @@
+    # column, we expect the func to have takern care of its own children.
+    f1.transform(replace_expr)
+    iterate = list(f1)
+    expected = [f1, c1, f2, f4, al1, f5]
```
Is there a reason you need to flatten the expected results rather than assert on the expected value after transformation?

fpacifici replied:
yes, f1.transform, transforms f1 in place:
the assertion would be assert f1 == f1

fpacifici committed:
Collapse Expressions and Conditions

fpacifici commented:
I turned Conditions into Expressions. This makes iteration and manipulation of the query a lot simpler: we do not have separate ConditionContainer and ExpressionContainer or wrapper classes in conditions to extract expressions.
The downside is that the type structure of the query is more permissive and it allows some clauses we may not want (even though they may be viable for clickhouse) like:
> SELECT (a = b) as something

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,161 @@
+from snuba.query.nodes import Node
+
+
+class Operator(Enum):
```
> a = SimpleCondition(col, Operator.IS, Null())

lynnagara replied:
Will we need "IS NOT" then?

fpacifici replied:
Not an issue anymore. Now we use functions for expressions

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,161 @@
+    NEQ = "!="
+    IN = "IN"
+    NOT_IN = "NOT IN"
+    IS = "IS"
```
`IS` is required for expresisons with NULL.
> col = NULL
generally does not work.
> col IS NULL
does

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,165 @@
+        """
+        Traverse the subtree in a prefix order.
+        """
+        yield self
```
yes, because iterating over a function would never yield the function see the result of the iteration in the test cases.
> test_expressions::test_iterate

fpacifici commented on tests/query/test_expressions.py:
```diff
@@ -0,0 +1,131 @@
+def test_iterate() -> None:
+    """
+    Test iteration over a subtree. The subtree is a function call in the form
+    f2(c3, f1(c1, c2))
```
damn I forgot about that syntax. Yes we can do that.

fpacifici commented on tests/query/test_expressions.py:
```diff
@@ -0,0 +1,131 @@
+def test_mapping_complex_expression() -> None:
+    """
+    Maps over an Expression container:
+    f(c1, fB(f())) -> f(c1, fB(f(f())))
```
because transform is not like a `map` operation. It updates the content of an expression container in place. This makes things easier when running the transform over a tree where it has to be called recursively:
a node `a` updates its children -> the parent node is supposed to update `a`.

fpacifici commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,161 @@
+                for sub in c:
+                    yield sub
+        else:
+            yield c
```
this code does not exist anymore

fpacifici committed:
Fix types

fpacifici committed:
Add BinaryCondition

tkaemming commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,43 @@
+    A container of nodes in the query.
+    This is an Iterable so we can traverse the tree and it provides
+    a map method to run a map on the content of the collection in
+    place.
```
This part of the comment about `map` is out of date.

tkaemming commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,43 @@
+
+        The solution is that we can call map only on containers and
+        they will transparently apply map on their content, but will
+        not produce a new container to replace the original one.
```
Similar comment here as above: this wording is out of date.

tkaemming commented on snuba/query/conditions.py:
```diff
@@ -0,0 +1,91 @@
+from snuba.query.expressions import Expression, ExpressionContainer
+
+
+class Operator(Enum):
```
Would it be simpler (on the internal query data structure side — not the parsing side) to just turn all of these infix operators into their functional equivalents, and remove the special casing for `Condition` types?

I'd think that this would be beneficial for any query processor to not have to handle both cases, and just expect everything to be functional. (I think the externally-facing query language should accept both.)

fpacifici replied:
That's a good question, I did not think about that assuming (probably incorrectly) that not all conditions could be expressed with functional equivalents.

I need to think about it for a second:
- yes it would make the structure simpler. 
- I need to verify whether query processing gains anything from explicit conditions.

fpacifici replied:
Turned Conditions into FunctionCall. This makes the conditions.py file simpler.
Right now it only states the valid function names, since the query processing code will have to depend  on those names to identify conditions, it is better not store these function names somewhere.

In the next step when I will introduce format and parsing, I am almost sure, I will reintroduce Some sort of Condition class. This time it will inherit from FunctionCall. This will allow us to format condition in the infix form (even when managed as functions) and it will simplify query processing (rely on class hierarchy instead of a string field name to identify the type of condition).
All this is not needed at this point, so I will bring it back later.

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Turn Conditions into FunctionCall

fpacifici committed:
Rename Conditions class

tkaemming commented on snuba/query/query.py:
```diff
@@ -47,11 +48,42 @@ class Query:
+        condition: Optional[Expression] = None,
+        groupby: Optional[Sequence[Expression]] = None,
+        having: Optional[Expression] = None,
+        order_by: Optional[Sequence[Expression]] = None,
```
Shouldn't expressions in the order by have a direction associated with them?

fpacifici replied:
That's a good point. I forgot about that.

tkaemming commented on snuba/query/query.py:
```diff
@@ -47,11 +48,42 @@ class Query:
+        data_source: RelationalSource,
+        # New data model to replace the one based on the dictionary
+        selected_columns: Optional[Sequence[Expression]] = None,
+        aggregations: Optional[Sequence[NodeAggregation]] = None,
```
Would it make sense to try to consolidate them as part of this process and only introduce aggregations as a distinct type if there is a roadblock that's not worth tackling as part of the migration process? Or is there one that you already know for sure that you'll run into?

fpacifici replied:
I don;t think aggregations should exist on their own, but, as of today they are managed in a very different way by column_expr with respect to other expressions (don't know what the reason is), So preserving that would make it the rollout much easier since I will be able to compare the result of the parsing of this data model with the old query structure.
About roadbloacks, I am not aware of any, but I would not be surprised there were. I added them to increase confidence during the rollout.

tkaemming commented on snuba/query/query.py:
```diff
@@ -47,11 +48,42 @@ class Query:
+        # New data model to replace the one based on the dictionary
+        selected_columns: Optional[Sequence[Expression]] = None,
+        aggregations: Optional[Sequence[NodeAggregation]] = None,
+        array_join: Optional[Column] = None,
```
Do you know if this is actually used anywhere? (If not, can we remove it?)

In my opinion, we should try to deprecate this long term and just replace it with a generic join keyword.

fpacifici replied:
I did not check, can verify. Actually removing it will be very convenient if possible.

fpacifici replied:
Unfortunately it is used for discoverV2. I need to dig deeper to understand where

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,183 @@
+from snuba.query.collections import NodeContainer
+
+
+class Expression(ABC):
```
Not specific to this line, but I think this is missing literal subtypes of `Expression` for string, number, array, etc.

fpacifici replied:
right.

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,183 @@
+
+
+@dataclass
+class AliasedExpression(Expression, ExpressionContainer):
```
I think ClickHouse might allow you to alias *any* arbitrary expression, so I'm wondering if adding the alias might be better on the expression after all:

```
SELECT 
    a, 
    b, 
    c
FROM system.one 
GROUP BY dummy AS a
HAVING a = 0 AS b
ORDER BY b AS c ASC

┌─a─┬─b─┬─c─┐
│ 0 │ 1 │ 1 │
└───┴───┴───┘

1 rows in set. Elapsed: 0.013 sec. 
```

I'm not entirely sure how all of these aliases are resolved though, so not sure how much sense this makes generically.

fpacifici replied:
In the current abstraction, where everything ended up being an expression, I don't think it makes much difference. The only trick is manipulating aliases, which is marginally simpler if the alias is an attribute: 
Imagine this case: I have an expression like this f(A) AS B and I want to transform it into g(x) as B.

If the alias is an attribute:
I will have a node f(A) as B and I just replace it with a node g(x) with the same alias. Simple

If the alias is a Wrapper:
I would have a node f(A) wrapped by an Alias B. Now When I transform f(A) into g(X) I do not have access to the alias since it is a parent node, which means I will end into an Alias node A wrapping an alias node with alias f(A) which wraps G(X) (if we replace node A we swap it with node B with alias A to preserve the naming). 
Which means at the end of the process we need to collapse aliases. It is feasible and it would work, it is just a little more complex.

tkaemming reviewed (commented):
I still want to put some thought into now `NodeContainer` and `ExpressionContainer` work here — something about them seems off to me, but overall I think that this is moving in the right direction.

I put some other more specific comments inline.

fpacifici commented:
_I still want to put some thought into now NodeContainer and ExpressionContainer work here — something about them seems off to me, but overall I think that this is moving in the right direction._

Indeed, the NodeContainer interface comes from a time when conditions and expressions were separate and we needed to replace expressions included in a condition.  Now that some of those requirements are gone, it may be possible to revisit this. At least the transformation method. I think the iterable mechanism is fine. What do you think ?

fpacifici committed:
Turn alias back into an attribute of an expression

fpacifici committed:
Turn transform in place into a mapping function

fpacifici committed:
Add literals

fpacifici committed:
Add ORderBy expression

fpacifici committed:
Implement the iterator on the query class

fpacifici commented:
Changes since last review:
- revisited the concept of NodeContainer and went back to the original idea of a proper map function that transforms and expressions and return the new value. Now, all nodes are expression, so this can work.
- changed the way we traverse the tree from prefix to postfix to accomodate the semantics of the transform function (see comments in line)
- added a class to manage Literals and OrderBy direction
- Rewrote the documentation to follow the new design.

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    A literal in the SQL expression
+    """
+    value: str
```
I suspect that this will end up being a union type that includes other types (e.g. numerics.)

Is there any reason to make a distinction between `Literal(None)` and `Null()`? That seems redundant at this point.

fpacifici replied:
No there is not. Will merge

fpacifici replied:
I don't think we need the sequence of literals, because, like other conditions it will be expressed as a function:

```
5c4070ddfe70 :) SELECT event_id from sentry_local where in(project_id, tuple(2, 3));

SELECT event_id
FROM sentry_local
WHERE project_id IN (2, 3)

Ok.
```

Moreover in() accepts expressions. Not that it would be a good idea to do that.
I think we should add meaningful subclasses of FunctionCall when we start using it to reduce boilerplate.

tkaemming replied:
Good point — I hadn't considered this, I guess as long as we have all the scalar types, we should be able to use `array` and/or `tuple` to do everything else we need.

Looking at the [ClickHouse data types documentation](https://clickhouse.yandex/docs/en/data_types/) and considering our usage, we might want to consider adding UUID and Date/DateTime types? I'm not sure how these would be handled at the parser level to disambiguate them from plain strings, though…

fpacifici replied:
Unfortunately, the UUID comes as a string from the client, the parser does not know whether it should be translated into a clickhouse UUID or to be kept as a string. Datetime can be managed.

tkaemming replied:
I wonder if it'd make sense (at some point — not necessarily now) to make `value` a type variable so that the Snuba query and the ClickHouse (or whatever other table engine, per #582) query have different possible sets of literals. I can't think of any other ways that I'd expect the two different query structures to diverge significantly, at least at this point.

Thinking about it more, I'm not sure that introducing a DateTime type is a good thing here. I'd assume we'd have to do so using regular expressions?

fpacifici replied:
What about we wait when we start building some actual query processing to see what level of specialization makes sense for literals? I am not sure we should even have DateTime. There is no way in the query language to tell when a literal is actually a DateTime or a string containing a Datetime.

tkaemming replied:
I think that is fine.

> I am not sure we should even have DateTime. There is no way in the query language to tell when a literal is actually a DateTime or a string containing a Datetime.

For what it's worth, I think we're on the same page here.

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,217 @@
+        its own way.
+
+        It returns the new expression. It does not guarantee the returned value
+        will be a new object. No guarantee of immutability is given on expressions.
```
Let me break the comment into two:
1) "It does not guarantee the returned value will be a new object." This is inevitable since it depends on what the function passed does. On a leaf node, if `func` returns self, it is the same object. If it builds a different object instead of mutating the existing then it is a new object.
2) "No guarantee of immutability is given on expressions." If we do not impose immutability, the point at number 1 implies that a transform may return the current object mutated or a new one, we cannot do anything about that. We can try to force all expressions to be immutable. This would be trivial for leaves, but more complex for expresisons that have children since children are a sequence. In order to make HierarchicalExpression immutable, we need to be able to tell either that the transformation of the children did not produce a change, thus we can return self (feasible) or always assume that we would instantiate a separate HierarchicalExpression every time transform is called on it, whether the content was changed or not, which does not seem very efficient. 

Does it make sense ?

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,217 @@
+        raise NotImplementedError
+
+
+class HierarchicalExpression(Expression, NodeContainer[Expression]):
```
The only reason was that I did not like the idea of iterating on a leaf node (which does not contain nodes). Being able to do this seemed conceptually a counterintuitive api:
```
c = Column("alias", "column", "table")
for col in c:
    yield c
```

This is the sole reason for not making expression a NodeContainer itself.

About the OrderBy, the problem would exist anyway but it is not visible at this stage (mypy is fine now) because it would present itself when transforming a container of OrderBy classes. 
```
container: Sequence[OrderBy] = []
# stuff
container = map(o.transform(func), container)
```
This will happen no matter what in a different place (when I want to apply transformation across the board on the query). There there are two possible solution I am still thinking over:
- Do not transform across the board but let the user decide which parts of the query to transform (so the sequence of order by will take e more specialized mapping function. This require ORderBy to override transform
- When transforming the expressions across the board only iterate over the content of OrderBy, which is an Expression. Which is probably what most use case will want to.

Probably the Query api will need to expose the two approaches.

tkaemming replied:
I'll come back to this one, but I think my thoughts here relate to the other comment that I just made above in a way that I haven't had time to clearly articulate yet.

I do think that `OrderBy` is not an expression because it's not substitutable with other `Expression` types.

fpacifici replied:
Now the hierarchical node still exists but it is not needed for correctness, it is just used to reuse logic between FunctionCall and Aggregation. Everything is an Iterable

fpacifici commented on snuba/query/query.py:
```diff
@@ -61,6 +94,35 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+        self.__condition: Optional[Expression] = condition
+        self.__groupby: Sequence[Expression] = groupby or []
+        self.__having: Optional[Expression] = having
+        self.__order_by: Sequence[OrderBy] = order_by or []
```
I think it can. I suspect I forgot them in place because in the first commit I did not pass the parameters to the constructor but I wanted the attributes to be created. Let me try to remove them.

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Make expressions immutable

fpacifici committed:
Make every expresison iterable

fpacifici committed:
Remove container

fpacifici committed:
Add a replacement method in the Query class

fpacifici commented:
Updates:
- now expressions are immutable, so transform works like a map method
- all expressions are now iterable. Which removes the NodeContainer
- use itertools.chain instead of CompositeNodeContainer
- add a transform method to Query and the relative test.

fpacifici commented on snuba/query/collections.py:
```diff
@@ -0,0 +1,42 @@
+        raise NotImplementedError
+
+
+class CompositeNodeContainer(NodeContainer[TNode]):
```
removed

fpacifici committed:
Fix iteration

fpacifici commented on snuba/query/query.py:
```diff
@@ -62,6 +96,58 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+            chain(*map(lambda orderby: orderby.node, self.__order_by)),
+        ])
+
+    def transform_expression(
```
See test_query_ast.py for a concrete example.

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,215 @@
+
+
+@dataclass(frozen=True)
+class Aggregation(HierarchicalExpression):
```
Would it make sense to make `Aggregation` a subtype of `FunctionCall`? Or do you think it's better to keep them siblings in the type hierarchy? Seems to me that might be the best of both worlds for both consistency as well as the ability to do `isinstance(expr, Aggregation)` for now.

This would also make `HierarchicalExpression` unnecessary (of which I still think removing would make the design easier to understand.)

tkaemming commented on snuba/query/query.py:
```diff
@@ -62,6 +96,58 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+        function to all expressions contained in this query
+        """
+
+        def transform_expression_list(expressions: Sequence[Expression]):
```
This needs a return type.

tkaemming commented on snuba/query/query.py:
```diff
@@ -62,6 +96,58 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+            )
+
+        self.__selected_columns = transform_expression_list(self.__selected_columns)
+        self.__aggregations = transform_expression_list(self.__aggregations)
```
Does this error in mypy? I'd assume it should, since this is assigning `Sequence[Expression]` to `Sequence[Aggregation]`.

fpacifici replied:
You are right, I need to make this function generic. I did not run a full mypy --strict after this last change.

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,215 @@
+    processing still relies on aggregation being a first class concept.
+    """
+    function_name: str
+    parameters: Sequence[Expression]
```
Do you have any initial thoughts on how this will support parametric functions (e.g. `topK(5)(col)`)?

fpacifici replied:
It seems it should be a subclass of FunctionCall and I was planning to address that when needed to reduce the size here. But I can add it now if you prefer.

tkaemming replied:
I'll leave that up to you as long as there's a plan.

tkaemming commented on tests/query/test_query_ast.py:
```diff
@@ -0,0 +1,104 @@
+        column1, agg1,
+    ]
+
+    assert list(query.get_all_expressions()) == expected_expressions
```
Can this assertion be rewritten to compare a `Query` to a `Query` without the indirection of `get_all_expressions`? `get_all_expressions` is already tested, and I feel like flattening out the query like this makes it difficult to understand and also doesn't ensure that the transformation function preserves the query tree structure.

fpacifici replied:
Did not try but I guess it should work.

fpacifici commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,215 @@
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
```
I am still trying to figure out the way to format, and I need to do some additional experiments before having a strong opinion. Indeed the idea you are proposing is one of the three approaches:
1) first transform the query into a ClickhosueQuery, that would be structured, and then having a format method on that.
2) the clickhouse module could have a format method that takes a Query, it navigates it and produces the string output. This is simple to understand but the navigation logic becomes complex since the clickhouse query would need to know the internals of the Query object
3) Use a _strategy pattern_, which means passing a ClickhouseFormatter object to the format method in the Query class (which is what I wanted to experiment first). ClickhosueFormatter provides low level primitives (like how to escape columns, how to format a JOIN statement and so on), but the Query object retains the responsibility of traversing the tree, that can be done recursively in a very easy way then. 

Now that we are landing onto the idea that we will have to process a structured clickhouse query as well, I think I should move from 3 to 1.
I am ok in removing the format method for now (anyway it is not implemented at all, it does not change the content of this PR). What do you think ?

tkaemming replied:
> Now that we are landing onto the idea that we will have to process a structured clickhouse query as well, I think I should move from 3 to 1. I am ok in removing the format method for now (anyway it is not implemented at all, it does not change the content of this PR). What do you think ?

I agree with both of these points. I think #574 highlighted the need for a separate query object because there are clauses like `PREWHERE` in the ClickHouse query that aren't clauses in the Snuba query. I'd expect that we could reuse a lot of the typing work done here in that class, though.

I think we could figure out how to do one of the other two options if we needed to, but I think it'd be easier to understand (and test) things like prewhere optimization by separating that concern from other ones like formatting.

fpacifici replied:
If 3 does not show any evident blocker I am not going to do neither 1 nor 2

tkaemming replied:
I'm confused — are we back where we started now? Or something else…?

fpacifici replied:
Sorry, I messed up the numbers. Natural numbers are hard, I prefer complex ones.
if **1** does not show any evident blocker I am not going to do neither **2** nor **3**

tkaemming replied:
Ok, that makes more sense. 😄

fpacifici commented on snuba/query/query.py:
```diff
@@ -62,6 +96,58 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
         self.__prewhere_conditions: Sequence[Condition] = []
 
+        # New data model
+        # TODO: Provide a better typing for this.
```
no, will remove

tkaemming replied:
Ok — one more change (sorry): this line was never removed.

tkaemming replied:
Please don't forget about this.

fpacifici replied:
That's weird I was sure I removed it. I guess it must have came back during the merge for black.

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Review comments

fpacifici committed:
Blast Aggregation out of existence and live a happy life as a consequence

fpacifici committed:
Check the structure of the query in the test

fpacifici committed:
Some type fixes

fpacifici commented:
Updates:
- Aggregations pissed me off enough to make it type check properly that I just removed it. Introducing a structured clickhosue query should help in testing the old and the new formatted results are compatible
- Support multi parameter group function class ( f(X)(Y) )
- Add more getters for the fields in the Query (now used only for assertions in tests)
- remove format methods
- some type fixes

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,132 @@
+    parameters_group1: Sequence[Expression]
+    # This is for expressions like f(x)(y).
+    # None means there is no second gorup of parameters
+    parameters_group2: Optional[Sequence[Expression]] = None
```
Would it make sense to instead structure this as:

```
function: Union[str, FunctionCall]
parameters: Sequence[Expression]
```

… or is this too permissive, since this could lead to calls like `f(g(x)(y))(z)`? (Is that even a valid construct? I haven't tested it yet, so I honestly have no idea.)

fpacifici replied:
I think the issue with this is not the example you gave. That one is valid, but it was possible before as well by having one parameter of the first group of f, being a FunctionCall.
The issue is that I could express this:
```
f(x)(y)(z)(y)
```

_My functional programming enthusiast self speaks now:_
My interpretation of this syntax in Clickhouse is that that two sets of parameters are just useless syntactic sugar. But now that you mention it, this may actually a proper currying implementation, in which case it would make sense to consider the external call as an application of a function to the function returned by the application of the internal one.
z = f(x)(y) = f(x) returns g(k) -> z = g(y).
Which means f(x)(y)(z)(w) should be allowed. 
_My functional programming enthusiast self stops speaking now_

In practice my concern with expressing currying properly is that, while it is extremely powerful to evaluate expressions, it may make query processing harder. 
I think it is worth giving it a shot anyway, I may just split the FunctionCall implementation, instead of using the Union, this will make it clearer for query processors. 

```
# I need to think over naming better. But this is the idea.
class FunctionCall(ABC):

class BasicFunctionCall(FunctionCall):
    name: str
    params[...]

class CurriedFunctionCall(FunctionCall):
    internal: FunctionCall
     params[...]
```

What do you think?

tkaemming replied:
> I think the issue with this is not the example you gave. That one is valid, but it was possible before as well by having one parameter of the first group of f, being a FunctionCall.

Oops — yeah, that was a bad example, but based on the remainder of your comment I think you ended up on the same train of thought that I was.

> But now that you mention it, this may actually a proper currying implementation, in which case it would make sense to consider the external call as an application of a function to the function returned by the application of the internal one.

Yeah, this is how I've thought always of it — whether or not that's an accurate reflection of their intent, I'm not sure one way or another.

> In practice my concern with expressing currying properly is that, while it is extremely powerful to evaluate expressions, it may make query processing harder.

Do you have any examples why? Nothing sticks out to me immediately but I haven't put much thought into this.

> I think it is worth giving it a shot anyway, I may just split the FunctionCall implementation, instead of using the Union, this will make it clearer for query processors.
> 
> ```
> # I need to think over naming better. But this is the idea.
> class FunctionCall(ABC):
> 
> class BasicFunctionCall(FunctionCall):
>     name: str
>     params[...]
> 
> class CurriedFunctionCall(FunctionCall):
>     internal: FunctionCall
>      params[...]
> ```
> 
> What do you think?

I think this makes sense in general.

I don't think that we need a shared supertype for these — I'd consider just having `FunctionCall(Expression)` and `CurriedFunctionCall(Expression)` with the same attributes you described above. I think that not allowing infinitely recursive currying would be an advantage, since it's unclear whether or not ClickHouse allows that (and I don't think we have a use-case for it right now.)

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,132 @@
+        Transforms the subtree starting from the children and then applying
+        the transformation function to the root.
+        This order is chosen to make the semantics of transform more meaningful,
+        the transform operation will be performed on thechildren first (think
```
```suggestion
        the transform operation will be performed on the children first (think
```

tkaemming commented on tests/query/test_query_ast.py:
```diff
@@ -0,0 +1,107 @@
+        groupby=[FunctionCall("alias", "tag", [Literal(None, "f1")])],
+        having=None,
+        order_by=[orderby],
+    )
```
This is much nicer, thank you.

tkaemming commented on snuba/query/query.py:
```diff
@@ -35,6 +40,23 @@
+        """
+        Returns a new OrderBy clause with a new node.
+        """
+        return OrderBy(self.direction, new_node)
```
Can callers use [`dataclass.replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) instead of this function, or does that have typing implications? Looking at the signature of `replace` I could see how some type information would be lost (but I haven't tested this.)

I general I think actually exposing a `replace` method here (and on `Expression`) that binds the instance argument of `replace` to `self` would be the nicest API from a caller's perspective. I suspect that this pattern will be used a lot in `transform` functions.

fpacifici replied:
For now I relied on dataclasses.replace because, if I defined the replace method in the classes themselves, I did not manage to convince mypy that the types made sense.

tkaemming commented on snuba/query/query.py:
```diff
@@ -62,6 +114,57 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+            chain(*self.__groupby),
+            self.__having if self.__having else [],
+            chain(*map(lambda orderby: orderby.node, self.__order_by)),
+        ])
```
`chain(*[x, y, z])` is redundant, this should probably just be `chain(x, y, z)`.

tkaemming commented on snuba/query/query.py:
```diff
@@ -62,6 +114,57 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+        tree.
+        """
+        return chain(*[
+            chain(*self.__selected_columns),
```
Would `chain.from_iterable` work here instead of argument unpacking?

tkaemming commented on snuba/query/query.py:
```diff
@@ -62,6 +114,57 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+            chain(*map(lambda orderby: orderby.node, self.__order_by)),
+        ])
+
+    def transform_expression(
```
This should probably be pluralized (`transform_expressions`) but I think `transform` might be better for consistency, unless you think that is going to be confusing.

tkaemming commented on snuba/query/query.py:
```diff
@@ -79,6 +182,9 @@ def __extend_sequence(self,
     def get_selected_columns(self) -> Optional[Sequence[Any]]:
         return self.__body.get("selected_columns")
 
+    def get_selected_columns_exp(self) -> Sequence[Expression]:
```
This suffix is kind of misleading — this isn't an expression, it's a sequence of expressions. `_exprs` (or even `_expressions`) is probably more appropriate here (and some of the methods below.)

tkaemming commented on snuba/query/query.py:
```diff
@@ -151,9 +263,15 @@ def get_arrayjoin(self) -> Optional[str]:
     def get_orderby(self) -> Optional[Sequence[Any]]:
         return self.__body.get("orderby")
 
+    def get_orderby_exp(self) -> Sequence[OrderBy]:
```
`_exp` doesn't make a lot of sense here (since order by is not an expression.) I'm not exactly sure what to call these though — they're not clauses, either. 😕

fpacifici replied:
To me these methods will be renamed as get_orderby, get_conditions, etc. as soon as the AST is in prod and the old system based on the dictionary is  removed.

fpacifici committed:
Use dataclasses.replace to duplicate expressions

fpacifici committed:
Merge branch 'master' into feat/aColumnAbstraction

fpacifici committed:
Fix indentation

fpacifici committed:
Nits

fpacifici committed:
Fix types

fpacifici committed:
Split FunctionCall from CurriedFunctionCall

fpacifici committed:
Some more nits

fpacifici commented:
Changes:
- mostly cosmetics
- CurriedFunctions have their own class.

tkaemming commented on snuba/query/expressions.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        transformed = replace(
+            self,
+            internal=func(self.internal_function),
```
```suggestion
            internal_function=func(self.internal_function),
```

tkaemming commented on snuba/query/query.py:
```diff
@@ -57,6 +104,57 @@ def __init__(self, body: MutableMapping[str, Any], data_source: RelationalSource
+        """
+        Transforms in place the current query object by applying a transformation
+        function to all expressions contained in this query
+        """
```
Should this return a new `Query` to mirror the semantics of `transform`?

fpacifici replied:
I wish , unfortunately we cannot do that just yet:
There are several places during the query execution that rely on Request being a sealed dataclass with a query argument, thus We cannot swap the Query as of now. 
This is what drove the QueryProcessor to make changes to the query object in place.

We need first to remove all dependencies on a mutable Query, then we can make Query immutable itself.

tkaemming replied:
That makes sense. It'd probably a good thing to note in a comment that this was a design decision made as a side effect of another design decision (so we don't have to remember why it was written this way if/when we can change it later.)

tkaemming reviewed (changes requested):
One typo and one question/possible change, otherwise this looks good to me. Thanks for working so hard on this!

fpacifici committed:
Fix transformation on curried functions

fpacifici committed:
Add comment about Query mutability

fpacifici committed:
Remove old comment
